Adding WB hashing for bls12-381 and bls12-377

drskalman opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

drskalman committed:
include the coefficients of bls12-377 G1 and G2 isogenies for wb hash2curve

drskalman committed:
replacing the ark-ec dependency to w3f/arkworks-algebra

drskalman committed:
added parameters for isogenous curve to bls12-377 g1 curve for swu map

drskalman committed:
fix dependency inconsistency problem

drskalman committed:
implement WBParams and its isogenous curve for BLS12-377 G1

drskalman committed:
implement the SWUParameters for isogenous curve to BLS12-377 g2 curve.

drskalman committed:
implement WBParams for BLS12-377 G2 curve

drskalman committed:
replacing const generic arrays with const slice in g1 and g2 iso coefficients

drskalman committed:
- Implement WB hash to curve for BLS12-381 G1
- Improvement to WB hash to curve code for BLS12-377
- Update sage code for generating isogeny coefficients for arkworks.

drskalman committed:
Implement WB hash to curve for BLS12-381 G2

drskalman committed:
- fix the bug with base order for field_new for Fq2
- fix the polynomial coeffcient order for bls12-381 g1 and g2 isogenies.
- fix the polynomial coeffcient order for bls12-377 g1.
377-g2 needs to be fixed still.

drskalman committed:
fix bls12-377 wb hash to curve isogeny coeffcients

drskalman committed:
fixed sage code for generating the isogeny coefficients

drskalman committed:
Merge branch 'master' into skalman-bls12-377-wb-hash2curve

drskalman committed:
Merge remote-tracking branch 'origin/master' into skalman-bls12-377-wb-hash2curve

drskalman committed:
use patch.crates-io hack to resolve dependancy issues on w3f fork

Pratyush committed:
Rename `Parameters` to `Config` for all fields

Pratyush committed:
Rename `field_new` to `MontFp`

Pratyush committed:
Rename `field_new` to `QuadExt` and `CubicExt`

drskalman committed:
- Merge remote-tracking branch 'upstream/upgrade-ff' into skalman-bls12-377-wb-hash2curve
- make bls12_377 implementation compatible with the upgrade.
- bls12_377 builds

drskalman committed:
Refactor bls12_381 crate based on update-ff branch. Close #9

drskalman committed:
Merge branch 'master' into skalman-bls12-377-wb-hash2curve

drskalman committed:
adapt bls12 wb hashing to new changes in algebra

drskalman committed:
update bls12-381 dependancies and g1, g2 definition to pass tests

drskalman committed:
adapt bls12-377 hash to curve to algebar updates

drskalman committed:
depend on upstream for pull request

drskalman committed:
cargo fmt

drskalman committed:
- move the isogeny finder script from sage to script folder
- delete auxiliary isogeny coeff file

drskalman committed:
add unit tests for wb hashing to bls12-377 g1 and g2

mmagician reviewed (commented):
Looks good 👍🏼

Pratyush commented on bls12_377/src/curves/g2_swu_iso.rs:
```diff
@@ -0,0 +1,118 @@
+// 8444461749428370424248824938781546531375899335154063827935233455917409239041
+//(0 : 1 : 0)
+/// G2_GENERATOR_X_C0 =
+pub const G2_GENERATOR_X_C0: Fq = MontFp!("44471777796618567688228760095584248343372454885978087674329841655595593880133139294404651664057692271364857231527");
+
+/// G2_GENERATOR_X_C1 =
+pub const G2_GENERATOR_X_C1: Fq = MontFp!("152209914092745808277594956866181055187624831129109767937025242463317365117655129123148193049673425418513926319001");
+
+/// G2_GENERATOR_Y_C0 =
+pub const G2_GENERATOR_Y_C0: Fq = MontFp!("115206687171448860889110309021279060303629519187879257051215751573842462972180856243991157572371972099444077110343");
+
+/// G2_GENERATOR_Y_C1 =
+pub const G2_GENERATOR_Y_C1: Fq = MontFp!("191377956145194479040228903677940355038998863371661730030204479850936075480341608934735952709786495341106477933498");
```
Can we directly specify this inline in `G2_GENERATOR_X` and `G2_GENERATOR_Y`?

drskalman replied:
fixed [here](https://github.com/w3f/arkworks-curves/pull/11/commits/86ec5aae5708921f2040147d8c6fd1e30cc220e4#diff-5f6dece84d541d07eee0ae081e6e8b1c3ba876851c7591391f07bab9591122d0R74) and [here](https://github.com/w3f/arkworks-curves/pull/11/commits/86ec5aae5708921f2040147d8c6fd1e30cc220e4#diff-3658c7f4fe12cc2a0190ce2b0809d509a3bc57f5e9737869593afcc341e94f4cR68)

drskalman commented:
Pending https://github.com/arkworks-rs/algebra/pull/517

drskalman committed:
- Use IsogenyMap struct to specify WB Isogeny for bls12-381 and bls12-377
- Do not use auxiliary constants to define generators of g2_swu_iso curve.
- Update change log

drskalman committed:
Merge branch 'master' into skalman-bls12-377-wb-hash2curve

drskalman committed:
Bump the Algebra dependencies of bls12_381 and bls12_377 to 0.4.0-alpha.4 so they could use the IsogenyMap struct.

drskalman committed:
Drop alpha sub-version in dependancies because it takes the lastest sub version

drskalman commented:
I have @mmagician oral blessing to merge.

burdges commented:
IRTF H2C draft says Edwards curves should use Elligator, not WB.  In fact, this means, or should mean, curves for which an Edwards form exists, should use Elligator, not WB.  Ergo, people should really not be using WB for BLS12-377.
